implementations/rust/ockam/ockam_examples/example_projects/no_std/src/tracing_subscriber.rs

```diff
@@ -0,0 +1,235 @@
+/// A simple tracing_subscriber implementation for no_std targets
+use tracing::{
+    dispatcher::Dispatch,
+    field::{self, Field},
+    span::{Attributes, Record},
+    Event, Id, Metadata, Subscriber,
+};
+
+// - tracing_println ----------------------------------------------------------
+
+#[macro_export]
+macro_rules! tracing_println {
+    ($($arg:tt)*) => {{
+        // dummy logger
+        #[cfg(not(any(feature="log-itm", feature="log-semihosting", feature="log-uart")))]
+        {
+            use ockam_core::compat::io::Write;
+            let mut buffer = [0 as u8; 1];
+            let mut cursor = ockam_core::compat::io::Cursor::new(&mut buffer[..]);
+
+            match write!(&mut cursor, $($arg)*) {
+                Ok(()) => (),
+                Err(_) => (),
+            }
+        }
+
+        // itm logger
+        #[cfg(feature="log-itm")]
+        {
+            // give the itm buffer time to empty
+            cortex_m::asm::delay(350_000);
+            // print output using itm peripheral
+            let itm = unsafe { &mut *cortex_m::peripheral::ITM::ptr() };
+            cortex_m::iprintln!(&mut itm.stim[0], $($arg)*);
+        }
+
+        // semihosting logger
+        #[cfg(feature="log-semihosting")]
+        cortex_m_semihosting::hprintln!($($arg)*).unwrap();
+
+        // uart logger
+        #[cfg(all(feature="log-uart", feature="atsame54"))]
+        {
+            use atsame54_xpro as hal;
+            use hal::prelude::_embedded_hal_serial_Write;
+            cortex_m::asm::delay(500_000);
+            crate::uart_println!($($arg)*);
+        }
+    }};
+}
+
+// - registration -------------------------------------------------------------
+
+pub fn register() {
+    let subscriber = EmbeddedSubscriber::new();
+    let dispatch = Dispatch::new(subscriber);
+    tracing::dispatcher::set_global_default(dispatch)
+        .expect("global default dispatcher for tracing is already set");
+    tracing::debug!("tracing_subscriber::register initialized");
+}
+
+#[cfg(feature = "log-uart")]
+pub fn register_with_uart(uart: uart::Uart) {
+    unsafe {
+        uart::UART.replace(uart);
+    }
+
+    let subscriber = EmbeddedSubscriber::new();
+    let dispatch = Dispatch::new(subscriber);
+    tracing::dispatcher::set_global_default(dispatch)
+        .expect("global default dispatcher for tracing is already set");
+    tracing::debug!("tracing_subscriber::register_with_uart initialized");
+}
+
+// - EmbeddedSubscriber -------------------------------------------------------
+
+// TODO add a SubscriberBuilder so we can set log levels etc.
```

Comment:
Do you want to file an issue for this?

Reply:
If it's not worth doing til we need it, that's fine too. But we could possibly get a contributor to take it.

Reply:
I don't think we need to for this:

1. Any kind of EmbeddedSubscriber™ should never ship as part of an Ockam crate. The subscriber implementation for `tracing` depends on the specific board & other libs any given dev is targeting. So the `TODO` here is just a marker for anyone wanting to copy example code into their own project and wondering why their attempts to set log levels isn't working, it's not relevant to Ockam development.
2. This purpose of this commit is only to do the bare minimum to unblock @metaclips  last week so he can work on figuring out what we need to add testing against `qemu` to CI.
3. This example can be replaced soon anyway with a cleaner one I'm writing to test Embedded RNG stuff.